Run tests using GitHub Actions

samdoran opened the pull request:
Add tests that run in parallel using a matrix. Each test run corresponds to a `tox` environment. A custom container image is used for all tests except integration tests to improve test reliability and prevent having to install Python and `tox` for every test.

I haven't figure out how to run `podman` in a  Docker container so the integration tests run directly in the VM.

samdoran committed:
Update tox environments and ignore .tox dir

- add descriptions
- change default environment to run all tests
- create unit test environment

Update default env list

This is equivalent to the previous default job that skipped integration tests,
but now linters are also run by default.

webknjaz commented on tox.ini:
```diff
@@ -1,34 +1,32 @@
-commands =
-    py.test -v test -k "not integration" {posargs}
+    -r {toxinidir}/test/requirements.txt
+commands = pytest {posargs:test}
```
Instead of setting "test" here, it's usually better to set it in `test_paths =` in `pytest.ini`

samdoran replied:
Does that let you override it with positional args?

webknjaz replied:
Yep

samdoran committed:
Add CI jobs

Setup a test matrix for unit tests

samdoran committed:
Specify default test path in pytest.ini rather than tox.ini

samdoran committed:
Add step to save doc build artifacts

samdoran committed:
Run sanity tests before other tests

samdoran committed:
Build test images before running integration tests

The integration tests will use the local container rather than pulling it. It
might be a good idea to specify a different image name to ensure the locally
built image is used rather than transparently pulling the image if something
doesn't work right.

Shrews reviewed (commented):
Is there any advantage to having the docs job in GHA rather than Zuul? Losing the intermediate doc artifact to view the doc results is really a big hit here, and I'm not really seeing the advantage of moving that job. Is it a lot faster or something?

webknjaz commented:
Personally, I'd recommend enabling the PR integration in RTD. It integrates nicely with the PRs meaning that you get preview link in the checks every time and you don't need to go anywhere to find it.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,131 @@
+        uses: actions/checkout@v2
+
+      - name: Setup test environment
+        run: tox -e ${{ matrix.test.env }} --notest
```
Instead of `-e` in each place, consider just adding
```yaml
env:
    TOXENV: ${{ matrix.env }}
```
to this job. It's a widely used practice across the ecosystem.

samdoran replied:
That's a good idea.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,131 @@
+      - name: Run tests
+        run: tox -e ${{ matrix.test.env }}
+
+      - name: Save docs
```
You probably don't need this if you just enable RTD. It'll have PR previews out of the box, with no expiration.

samdoran replied:
I was trying to duplicate the functionality in Zuul that previews docs builds for PRs. I didn't know that was something that Read the Docs could do.

webknjaz replied:
Here's an example of how it looks in the checks section: https://github.com/ansible/pylibssh/pull/250.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,131 @@
   integration:
     runs-on: ubuntu-20.04
-    name: Integration
+    name: Integration - ${{ matrix.py_version.name }}
```
Note that the GH UI is quite narrow in some places. This is why verify that this fits and try to go for short job names.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,131 @@
-          python -m pip install --upgrade pip
-          pip install tox
+          ${{ format('python{0}', matrix.py_version.name) }} -m pip install --upgrade pip
+          ${{ format('python{0}', matrix.py_version.name) }} -m pip install tox
```
I wouldn't put this many separate commands in one step because it's hard to parse where one starts and the other one ends in the logs. Unless you add runner commands for collapsing the log groups around each.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,131 @@
-          pip install tox
+          ${{ format('python{0}', matrix.py_version.name) }} -m pip install --upgrade pip
+          ${{ format('python{0}', matrix.py_version.name) }} -m pip install tox
+          tox -e integration-${{ matrix.py_version.tox }} --notest
```
I don't understand why you'd need to explicitly pass the py version here. Can't you just set up the default python and use `TOXENV=integration-py` to use it?

samdoran replied:
Currently this is run in a VM, so I probably don't need to be explicit about the Python executable since the only Python version will be the one installed by the `setup-python@v2` action. I was mainly thinking about when I get these tests moved to running within the test container, which has multiple python versions installed. In that case `pip` is ambiguous.

But, inside the container image, `tox` is already installed, so I would be able to delete the `pip` commands.

In general, I prefer to be explicit about the Python executable that's invoking `pip` rather than just calling `pip` directly.

webknjaz replied:
What I observe across the ecosystem is that there's a `setup-python` step. Then, there's some `python -m pip install tox --user` (that uses the previously set up python interpreter). And then, there's `python -m tox --notest --skip-missing-interpreters false` (for provisioning) and `python -m tox --skip-missing-interpreters false` for running the actual testing.

And then, there would be a matrix of envs/pythons/OSs that makes up jobs that GHA would know better how to orchestrate.

webknjaz replied:
Why do you want to put this in a container, anyway?

webknjaz commented on test/unit/test_utils.py:
```diff
@@ -0,0 +1,104 @@
+
+
+@pytest.fixture
+def source_file(tmpdir):
+    source = os.path.join(tmpdir, 'bar.txt')
+    with open(source, 'w') as f:
+        f.write('foo\nbar\n')
+    return source
```
Why use tmp_dir?

```suggestion
def source_file(tmp_path):
    source = tmp_path / 'bar.txt'
    source.write_text('foo\nbar\n')
    return str(source)
```

samdoran replied:
I'm not sure I fully understand the question. Why use a temporary directory at all here? Are you saying this should just be a permanent test fixture rather than creating the same file each time? Or should I return a `Path` object if I'm going to bother changing this to `tmp_path`?

I didn't look at the broader use of the fixture and how it's used it tests, so there is most likely more room for improvement.

webknjaz replied:
This is mostly about the `tmpdir` fixture being deprecated while `tmp_path` is the thing that replaces it and should be always used instead.

samdoran replied:
Oh man, I had a total Monday moment. Sorry. I didn't realize that was a code suggestion. I thought it was a diff from my PR. 🤦‍♂️ 

I have switched to using `tmp_path` in _so many_ places I thought that was a change I had made here as well.

samdoran replied:
These fixtures are all original code I just moved. That's why they are showing up as "new" changes in this PR.

Ok, now I think I've un-confused myself. 😄

samdoran committed:
Add CI status badge

samdoran committed:
Add nightly runs

samdoran committed:
Remove docs artifact upload

This doesn't meet the goal of have previews of docs in PRs. This is something
that we may be able to do automatically using read the docs.

samdoran committed:
Use TOXENV environment variable rather than command line parameter

samdoran committed:
Separate tox installation and tox env setup steps

Also, don't be fancy about the Python executable.

Since integration tests are currently run in a VM which will only have one
Python version installed, I don't need to be explicit about the Python version
invoking pip.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,139 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        test:
+          - name: Lint
+            tox_env: linters
+
+          - name: Docs
+            tox_env: docs
```
looking at this: do you really need separate name and tox env? It's usually more than enough to reuse envs as job names:
```suggestion
        tox_env:
        - linters
        - docs
```

samdoran replied:
I could do that, I just like the way it looks having the job name capitalized. GHA doesn't provide an easy mechanism for manipulating the tox environment string into a name that I think looks nice, so that's the reason for separate entries. It's purely cosmetic.

webknjaz replied:
FWIW I'm working on a reusable workflow for tox and here's what it'll look like: https://github.com/webknjaz/tox-gha-test/actions/runs/1450588254 — with the exact env names.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,139 @@
+    needs: images
+
+    env:
+      TOXENV: ${{ matrix.py_version.tox_env }}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        py_version:
+          - name: 3.8
+            tox_env: py38
+
+          - name: 3.9
+            tox_env: py39
```
Once you select the current Python version, just using a static string `python` or `py` for `$TOXENV` is enough to get it picked up
```suggestion
      TOXENV: python

    strategy:
      fail-fast: false
      matrix:
        py_version:
        - 3.8
        - 3.9
```

samdoran replied:
> Once you select the current Python version

How do you "select" the Python version? By only installing one Python version with the `setup-python` action?

It seems like using `python` is ambiguous in an environment where multiple Python versions are installed, which is the case in the test image. Currently the integration tests run in the VM and I'm installing Python using the `setup-python` action. I would like to move away from that in the future once I figure out how to run Podman inside a Docker container.

In the test container image, all the Python versions are already installed as well as `tox`. I would think setting `TOXENV=python` in that case would pick up whichever Python version is sym linked to `/user/bin/python`, which may or may not be the version I want to test, unless there is some other tox magic I don't know about.

webknjaz replied:
`tox -e py` or `tox -e python` will select whatever currently is run if you just type in `python`. In GHA that would be whatever the last step running `setup-python` activated.

This is the approach you'd find in any (most?) of the big community projects (like setuptools, or tox, or pytest). It's been the de-facto approach for as long as I can remember. Definitely since the times Travis was popular, which then migrated to Azure and GHA.

Another approach you may see in such CI setups is running `tox --notest` first, for venv provisioning and then, running `tox` in a subsequent job. The reason is that it's nicer when you see separate steps in the CI output, it's more obvious which one failed (there was a bug in tox in some version that made people think that running tests failed but it was actually the pre-test stage, this was the case where the distraction was also useful). In some CIs, specifically Travis, there's different "stages". and depending on which stage errors out, it'd show different status for that job — error for the installation stage and failure for actual testing (or was it vice versa? I don't remember exactly) — this was another DX thingy that may be important.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,139 @@
+      matrix:
+        test:
+          - name: Unit
+            env: unit
```
this is unused

samdoran replied:
It seems it was used to create these entries in the test matrix:
 - [Unit - 3.8](https://github.com/ansible/ansible-builder/pull/284/checks?check_run_id=3793391339)  
 -  [Unit - 3.9](https://github.com/ansible/ansible-builder/pull/284/checks?check_run_id=3793391405)

samdoran committed:
Don't run all the tests for each matrix item

samdoran committed:
Split up integration tests

Run container builds and actual integration tests in parallel in order to
reduce overall test run time.

samdoran committed:
Remove external commands that are no longer required

samdoran committed:
Do not create junitxml file in test/artifacts

I'm not sure what this was used for and it deposits an artifact in the source
tree that is not ignored.

Shrews commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,135 @@
   pull_request:
-    branches: [devel]
+    branches:
+      - devel
```
I think if you just leave off the `branches` part, it will trigger for a pull request on _any_ branch, which would be what we want for backports to be tested properly.

samdoran replied:
This evaluates the base branch, not the source branch name. Limiting this workflow to only devel and release branches would prevent it  from running against other branches that get pushed to the repo. There shouldn't be many (any?) of those, though.

From the [GHA docs](https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#onpushpull_requestbranchestags):
> For a pull_request event, only branches and tags on the base are evaluated.

webknjaz replied:
Yep, Sam is correct: the `pull_request` event happens whenever there's an interaction with a PR. If there's a branch pushed to the upstream repo, it'll generate a `push` event, and only after a PR gets created there's *additionally/separately* a `pull_request` happening. If somebody would want to send a PR against a non-matching branch, this would not cause GHA to react to this specific event if the branch is not included.
Although, it's hard to imagine this happening often (or ever) so I don't see a strong argument for maintaining the list of branches for this event.

samdoran replied:
I'll remove the branch list.

Shrews commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,135 @@
+      - devel
+
+  schedule:
+    - cron: '00 04 * * *'
```
Not sure I see the need for scheduled CI runs. Is there a benefit I'm not seeing?

samdoran replied:
There's no real _need_ for it. I just figured it would be good to have at least some regular CI runs happening. Maybe weekly instead of nightly?

Shrews replied:
My reading of the GHA docs seems to indicate that the person that added this would get notified about the run status. If that's correct, then you probably shouldn't be the one adding this. I could add it later.

webknjaz replied:
@Shrews it's usually used for nightlies, as Sam mentioned. There's one quirk, though. In the event a repo stops being active, GitHub disables the workflows with "cron". An unobvious side-effect is that it does disable the whole workflow, not just the `schedule` trigger. So if there's other event triggers in the same workflow, this will result in new jobs not being started if somebody sends in a contribution after the disabling happens.

webknjaz replied:
> My reading of the GHA docs seems to indicate that the person that added this would get notified about the run status. If that's correct, then you probably shouldn't be the one adding this. I could add it later.

IIRC that is true, the person who pushed it gets notified (or is it the author of the last commit on the branch?)

samdoran replied:
That is very interesting and true. I'll go ahead and remove it for now.

samdoran committed:
Also trigger for release branches

samdoran committed:
Remove branch list and schedule

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,127 @@
-
   push:
-    branches: [devel]
+
```
Did you intentionally add this new line?

samdoran replied:
Yes, I have two blank lines between sections in the document so I added another here to match.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,127 @@
----
```
Any reason you didn't like this?

samdoran replied:
I think the document start marker should only exist if it's being used or if there are directives above the marker. Since this is a single document with no directives, it's not adding any value.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,127 @@
+      matrix:
+        py_version:
+          - name: 3.8
+            tox_env: unit-py38
+
+          - name: 3.9
+            tox_env: unit-py39
```
have you considered just using
```suggestion
            tox_env: unit

          - name: 3.9
            tox_env: unit
```

samdoran replied:
But that may not use the correct Python version.

webknjaz replied:
Ah, right, you don't use the setup action. The typical way is to invoke `setup-python` and then tox. Then it'll choose the right one.

samdoran replied:
The container image where the test is running has multiple Python versions installed.

webknjaz replied:
Oh, I missed that it's a container. Makes sense, then.

webknjaz replied:
I couldn't see why it diverged from what I typically used to be seeing.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,127 @@
+        run: tox
+
+
+  images:
```
Since this job is needed by others, maybe put it above them so it's logically ordered?

samdoran replied:
That makes sense.

webknjaz commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,127 @@
+        uses: actions/checkout@v2
+
+      - name: Create tox environment
+        run: tox --notest
```
A common theme you may notice across the popular project CI setups is that they use `--skip-missing-interpreters false` in every tox invocation, as a sanity check.
```suggestion
        run: tox --skip-missing-interpreters false --notest
```

samdoran committed:
Put command on single line since it's now just a simple command

samdoran committed:
Move images job above the integration job

Since integration depends on images, it reads better having it this way.

webknjaz commented on tox.ini:
```diff
@@ -1,39 +1,33 @@
     yamllint -s .
 
-[testenv:integration]
+[testenv:unit{,-py38, -py39}]
```
Is the whitespace needed here?

samdoran replied:
I don't think so.

samdoran committed:
Remove whitespace between factors

samdoran committed:
Disable unreliable test

webknjaz commented on test/integration/test_build.py:
```diff
@@ -148,6 +148,7 @@ def test_has_pytz(self, cli, container_runtime, pytz):
         r = cli(f'{container_runtime} run --rm {ee_tag} pip3 show pytz')
         assert 'World timezone definitions, modern and historical' in r.stdout, r.stdout
 
+    @pytest.mark.skip("Concurrency issues plus output parsing is using the incorrect format")
```
@samdoran for future, it's better to use xfail per https://pganssle-talks.github.io/xfail-lightning/ + https://pypi.org/p/pytest-rerunfailures exists.

samdoran replied:
Ah, that's good to know. This test is intermittent plus poorly written. I didn't know about `pytest-rerunfailures`.

webknjaz replied:
If it's flaky, use `xfail` + `strict=False` unless you want retries.
